File: usaco/12_13/5_mar/bronze/assign/assign.py
# This is essentially a line-for-line translation of the Java solution
# It will only pass the first four test cases
# (the Java solution passes all of them)

with open('assign.in') as f:
    n, k = map(int, f.readline().split())
    rules = []
    for _ in range(k):
        rel, x, y = f.readline().split()
        rel = 1 if rel == 'S' else 0
        x, y = int(x) - 1, int(y) - 1
        rules.append((rel, x, y))

def is_valid(assignments):
    for rel, x, y in rules:
        if rel == 1:
            if assignments[x] != assignments[y]:
                return False
        else:
            if assignments[x] == assignments[y]:
                return False
    return True

def count_valid_assignments(assignments, pos):
    if pos == n:
        return is_valid(assignments)
    total = 0
    for i in range(3):
        assignments[pos] = i
        total += count_valid_assignments(assignments, pos+1)
    return total

with open('assign.out', 'w') as f:
    f.write(str(count_valid_assignments([0]*n, 0)) + '\n')

File: usaco/12_13/5_mar/bronze/assign/assign.java
import java.util.*;
import java.io.*;

public class assign {
    private static int n;
    private static int k;
    private static int[][] rules;
    
    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(new BufferedReader(new FileReader("assign.in")));
        n = in.nextInt();
        k = in.nextInt();
        rules = new int[k][3];
        for (int i = 0; i < k; i++) {
            rules[i][0] = in.next().charAt(0) == 'S' ? 1 : 0;
            rules[i][1] = in.nextInt() - 1;
            rules[i][2] = in.nextInt() - 1;
        }
        in.close();
        
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("assign.out")));
        out.println(countValidAssignments(new int[n], 0));
        out.close();
    }

    private static int countValidAssignments(int[] assignments, int pos) {
        if (pos == n) {
            return isValid(assignments);
        }

        int total = 0;
        for (int i = 0; i < 3; i++) {
            assignments[pos] = i;
            total += countValidAssignments(assignments, pos+1);
        }
        return total;
    }

    private static int isValid(int[] assignments) {
        for (int i = 0; i < k; i++) {
            if (rules[i][0] == 1) {
                if (assignments[rules[i][1]] != assignments[rules[i][2]]) {
                    return 0;
                }
            } else {
                if (assignments[rules[i][1]] == assignments[rules[i][2]]) {
                    return 0;
                }
            }
        }
        return 1;
    }
}